mesonbuild/modules/rust.py

```diff
@@ -147,9 +147,12 @@ def test(self, state: 'ModuleState', args: T.Tuple[str, BuildTarget], kwargs: 'F
         new_target_kwargs['install'] = False
         new_target_kwargs['dependencies'] = new_target_kwargs.get('dependencies', []) + kwargs['dependencies']
 
+        sources = T.cast('T.List[SourceOutputs]', base_target.sources.copy())
```

Comment:
Seems odd we need to cast this, but your first commit changes derived classes and not BuildTarget itself.

Reply:
Oh, this is wonderful internal logic, I see...

Reply:
BuildTarget is already right, its only the derived classes that are wrong.

https://github.com/mesonbuild/meson/blob/master/mesonbuild/build.py#L682-L690

Reply:
As for the cast, this is the `list is invariant` problem again.
```python
x: List[A | B] = [A()]
```
Will give a mypy error because you're trying to do covariant assignment. Which is super annoying but I don't know any good way to solve.